# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/MotazSameh/flyrank-ml-internship/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

In [1]:
from datasets import get_dataset_config_names

configs = get_dataset_config_names("FlyRank/internship-warehouse")

print(configs)

['dim_clients', 'dim_content', 'fact_content_daily_performance', 'fact_content_query_90d']


In [2]:
from datasets import load_dataset

daily = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance",
    streaming=True
)

print(daily)

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

IterableDatasetDict({
    train: IterableDataset({
        features: ['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events'],
        num_shards: 18
    })
})


In [3]:
stream = daily["train"]

print(stream.features)

{'report_date': Value('date32'), 'client_hash_id': Value('string'), 'content_hash_id': Value('string'), 'client_has_gsc': Value('bool'), 'client_has_ga4': Value('bool'), 'gsc_data_available': Value('bool'), 'ga4_data_available': Value('bool'), 'gsc_impressions': Value('int64'), 'gsc_clicks': Value('int64'), 'gsc_sum_position': Value('int64'), 'gsc_avg_position': Value('float64'), 'ga4_pageviews': Value('int64'), 'ga4_sessions': Value('int64'), 'ga4_users': Value('int64'), 'ga4_engaged_sessions': Value('int64'), 'ga4_total_engagement_sec': Value('int64'), 'sessions_organic': Value('int64'), 'sessions_direct': Value('int64'), 'sessions_referral': Value('int64'), 'sessions_social': Value('int64'), 'sessions_paid': Value('int64'), 'sessions_ai': Value('int64'), 'ai_chatgpt': Value('int64'), 'ai_perplexity': Value('int64'), 'ai_gemini': Value('int64'), 'ai_copilot': Value('int64'), 'ai_claude': Value('int64'), 'ai_meta': Value('int64'), 'ai_other': Value('int64'), 'scroll_events': Value('in

In [2]:
from huggingface_hub import hf_hub_download
import pandas as pd

march_file = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    filename="fact_content_daily_performance/month=2026-03/data_0.parquet"
)

df_march = pd.read_parquet(march_file)

print("Shape:", df_march.shape)
print("\nColumns:")
print(df_march.columns.tolist())

print("\nDate range:")
print(df_march["report_date"].min(), "to", df_march["report_date"].max())

print("\nFirst 5 rows:")
display(df_march.head())

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Shape: (9841378, 30)

Columns:
['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']

Date range:
2026-03-01 to 2026-03-31

First 5 rows:


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,None,20,0,67,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,None,1,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,None,125,1,616,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,None,7,0,28,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,None,11,0,25,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [3]:
from huggingface_hub import hf_hub_download
import pandas as pd

april_file = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset",
    filename="fact_content_daily_performance/month=2026-04/data_0.parquet"
)

df_april = pd.read_parquet(april_file)

print("Shape:", df_april.shape)
print("\nColumns:")
print(df_april.columns.tolist())

print("\nDate range:")
print(df_april["report_date"].min(), "to", df_april["report_date"].max())

print("\nFirst 5 rows:")
display(df_april.head())

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  134MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Shape: (10424730, 30)

Columns:
['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']

Date range:
2026-04-01 to 2026-04-30

First 5 rows:


,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-04-01,client_62f4a7e64f5e0096,content_143987cfdeaba4c0,True,False,True,None,9,0,493,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-04-01,client_62f4a7e64f5e0096,content_13a8105125458098,True,False,True,None,1,0,9,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-04-01,client_62f4a7e64f5e0096,content_6a887d56ab6c8362,True,False,False,None,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-04-01,client_62f4a7e64f5e0096,content_e2bd76be7eed690d,True,False,True,None,1,0,7,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-04-01,client_62f4a7e64f5e0096,content_ddbfb1907979759a,True,False,False,None,0,0,0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### 1. Unit of analysis + time window

**Unit of analysis:** One row represents the daily performance of one content item for one client.

**Time window:** For this contract, I will use March 2026 (`2026-03-01` to `2026-03-31`) as the development window. This is a mid-panel month, so it is preferred over the final June 2026 sample when developing the data contract and feature logic.


In [4]:
# Verify the grain:
# one row should represent one content item × client × report date

total_rows = len(df_march)

unique_grain = df_march[
    ["report_date", "client_hash_id", "content_hash_id"]
].drop_duplicates().shape[0]

print("Total rows:", total_rows)
print("Unique content × client × date rows:", unique_grain)
print("Duplicate rows at this grain:", total_rows - unique_grain)

Total rows: 9841378
Unique content × client × date rows: 9841378
Duplicate rows at this grain: 0


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

### 2. Fields: feature / label / context / excluded

**Features**

I will use five observable performance features:

* `gsc_impressions` — available because Search Console records impressions before the refresh decision.
* `gsc_clicks` — available because Search Console records clicks before the decision.
* `gsc_avg_position` — available because the page's observed search position is measured before the decision.
* `ga4_pageviews` — available because Analytics records pageviews before the decision.
* `ga4_sessions` — available because Analytics records sessions before the decision.

**Label / proxy**

For the first version of this lane, the outcome will be treated as a future-performance proxy rather than using a current-period outcome as a feature. The exact label definition will be finalized after the time windows are established.

**Context**

* `client_hash_id` — identifies the client associated with the observation.
* `content_hash_id` — identifies the content item.
* `report_date` — identifies when the observation was recorded.

**Excluded**

I will exclude any feature derived from the future outcome window or directly derived from the label. These fields would not be knowable at the decision moment and could create target leakage.


In [5]:
# Verification Query 2:
# Row count and date window for March 2026

print("Row count:", len(df_march))
print("Minimum report_date:", df_march["report_date"].min())
print("Maximum report_date:", df_march["report_date"].max())

Row count: 9841378
Minimum report_date: 2026-03-01
Maximum report_date: 2026-03-31


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [6]:
# Verification Query 3:
# Check data availability using IS TRUE logic

gsc_available = df_march[df_march["gsc_data_available"].eq(True)]
ga4_available = df_march[df_march["ga4_data_available"].eq(True)]

print("Total March rows:", len(df_march))

print("\nGSC available (IS TRUE):", len(gsc_available))
print("GA4 available (IS TRUE):", len(ga4_available))

print("\nGSC availability rate:",
      round(len(gsc_available) / len(df_march) * 100, 2), "%")

print("GA4 availability rate:",
      round(len(ga4_available) / len(df_march) * 100, 2), "%")

Total March rows: 9841378

GSC available (IS TRUE): 3611061
GA4 available (IS TRUE): 413966

GSC availability rate: 36.69 %
GA4 availability rate: 4.21 %


In [7]:
# Five candidate features for the content-refresh lane

feature_cols = [
    "gsc_impressions",
    "gsc_clicks",
    "gsc_avg_position",
    "ga4_pageviews",
    "ga4_sessions"
]

feature_frame = df_march[
    ["report_date", "client_hash_id", "content_hash_id"] + feature_cols
].head(10)

display(feature_frame)

,report_date,client_hash_id,content_hash_id,gsc_impressions,gsc_clicks,gsc_avg_position,ga4_pageviews,ga4_sessions
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,20,0,3.350000,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,1,0,0.000000,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,125,1,4.928000,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,7,0,4.000000,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,11,0,2.272727,NaN,NaN
5,2026-03-01,client_73cda7b4e4f265ea,content_36c36abc7650d7af,239,1,7.347280,NaN,NaN
6,2026-03-01,client_73cda7b4e4f265ea,content_a7da352b73b02668,191,0,7.832461,NaN,NaN
7,2026-03-01,client_73cda7b4e4f265ea,content_05434271b257bb68,55,0,3.272727,NaN,NaN
8,2026-03-01,client_73cda7b4e4f265ea,content_d056587ff7faca0c,77,0,5.636364,NaN,NaN
9,2026-03-01,client_73cda7b4e4f265ea,content_bfd1e41c2af250c8,2,0,4.500000,NaN,NaN


### Five features and when they are available

1. **gsc_impressions** — knowable at the decision moment because Search Console records the page's observed impressions before a refresh decision.

2. **gsc_clicks** — knowable at the decision moment because Search Console records observed clicks before the decision.

3. **gsc_avg_position** — knowable at the decision moment because the page's observed search position is available from Search Console.

4. **ga4_pageviews** — knowable at the decision moment when GA4 data is available for the client and page.

5. **ga4_sessions** — knowable at the decision moment when GA4 data is available for the client and page.


In [8]:
# ============================================================
# Leakage Trap
# ============================================================

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Use a small sample to keep memory usage low
sample_keys = (
    df_march[
        ["client_hash_id", "content_hash_id"]
    ]
    .drop_duplicates()
    .sample(n=5000, random_state=42)
)

# Keep only the sampled March rows
march_sample = df_march.merge(
    sample_keys,
    on=["client_hash_id", "content_hash_id"],
    how="inner"
)

# Keep only the sampled April rows
april_sample = df_april.merge(
    sample_keys,
    on=["client_hash_id", "content_hash_id"],
    how="inner"
)

print("March sample rows:", len(march_sample))
print("April matching rows:", len(april_sample))

March sample rows: 148744
April matching rows: 150000


In [9]:
# Aggregate only the small sampled subset

march_perf = (
    march_sample
    .groupby(["client_hash_id", "content_hash_id"])["gsc_impressions"]
    .sum()
    .rename("march_impressions")
)

april_perf = (
    april_sample
    .groupby(["client_hash_id", "content_hash_id"])["gsc_impressions"]
    .sum()
    .rename("april_impressions")
)

outcome = (
    march_perf
    .to_frame()
    .join(april_perf, how="left")
    .fillna(0)
)

# Future outcome
outcome["future_decline"] = (
    outcome["april_impressions"] < outcome["march_impressions"]
).astype(int)

print("Sampled content items:", len(outcome))

print("\nFuture decline distribution:")
print(outcome["future_decline"].value_counts())

Sampled content items: 5000

Future decline distribution:
future_decline
0    3332
1    1668
Name: count, dtype: int64


In [11]:
import numpy as np
features = [
    "gsc_impressions",
    "gsc_clicks",
    "gsc_avg_position"
]

march_features = (
    march_sample
    .groupby(["client_hash_id", "content_hash_id"])[features]
    .mean()
)

honest_df = (
    march_features
    .join(outcome["future_decline"], how="inner")
    .dropna()
)

X_honest = honest_df[features].replace(
    [np.inf, -np.inf], np.nan
).fillna(0)

y = honest_df["future_decline"]

honest_tree = DecisionTreeClassifier(
    max_depth=2,
    random_state=42
)

honest_tree.fit(X_honest, y)

honest_pred = honest_tree.predict(X_honest)

honest_accuracy = accuracy_score(y, honest_pred)

print("Honest model accuracy:",
      round(honest_accuracy, 3))

Honest model accuracy: 0.639


In [12]:
# ============================================================
# DELIBERATE LEAKAGE
# ============================================================

leaky_df = honest_df.join(
    outcome["april_impressions"],
    how="left"
)

leaky_features = features + ["april_impressions"]

X_leaky = leaky_df[leaky_features].replace(
    [np.inf, -np.inf], np.nan
).fillna(0)

leaky_tree = DecisionTreeClassifier(
    max_depth=2,
    random_state=42
)

leaky_tree.fit(X_leaky, y)

leaky_pred = leaky_tree.predict(X_leaky)

leaky_accuracy = accuracy_score(y, leaky_pred)

print("Honest model accuracy:",
      round(honest_accuracy, 3))

print("Leaky model accuracy:",
      round(leaky_accuracy, 3))


Honest model accuracy: 0.639
Leaky model accuracy: 0.673


In [13]:
# Remove the leaked feature.
# It must not be part of the final feature set.

final_features = features

print("Final features after removing leakage:")
print(final_features)

Final features after removing leakage:
['gsc_impressions', 'gsc_clicks', 'gsc_avg_position']


### Leakage trap

I used April impressions as a deliberate leaked feature because April is the future outcome period relative to the March decision point. The model was allowed to see information that would not have been available when the decision was made.

The higher score from the leaky model is therefore not evidence of better ML performance. It is evidence of target leakage.

I removed `april_impressions` and kept only March features that were observable at the decision moment.


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

### 4. Data limits

This data can support observed and directional analysis of content performance, but it has important limitations.

* **Uneven data availability:** In March 2026, GSC data was available for 36.69% of rows, while GA4 data was available for only 4.21% of rows. Therefore, GA4-based analysis may not represent the full dataset.

* **Limited time window:** This analysis uses March 2026 as a mid-panel development month. A single month cannot represent all seasonal or long-term changes in content performance.

* **No causal conclusions:** The data can show associations and directional patterns, but it cannot prove that a particular content change caused a performance change.

* **Future information must be excluded:** Features derived from a future outcome window cannot be used at the decision moment because they would create leakage.

* **Client-level variation:** Different clients may have different levels of data availability and different performance patterns, so results from one client or group of clients may not generalize to all clients.


In [14]:
# Data availability limitation

total_rows = len(df_march)

gsc_available = df_march["gsc_data_available"].eq(True).sum()
ga4_available = df_march["ga4_data_available"].eq(True).sum()

print("Total rows:", total_rows)
print(f"GSC available: {gsc_available:,} ({gsc_available / total_rows * 100:.2f}%)")
print(f"GA4 available: {ga4_available:,} ({ga4_available / total_rows * 100:.2f}%)")

print("\nLimitation:")
print("GA4 coverage is limited, so GA4-based features may not represent all rows.")

Total rows: 9841378
GSC available: 3,611,061 (36.69%)
GA4 available: 413,966 (4.21%)

Limitation:
GA4 coverage is limited, so GA4-based features may not represent all rows.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.